qualtran/bloqs/generalized_qsp.py

```diff
@@ -79,6 +79,114 @@ def decompose_from_registers(
+    d = len(P) - 1  # degree
+
+    # R(z) = z^d (1 - P^*(z) P(z))
+    # obtained on simplifying Eq. 34, Eq. 35 by substituting H, T.
+    # The definition of $R$ is given in the text from Eq.34 - Eq. 35,
+    # following the chain of definitions below:
+    #
+    #     $$
+    #     T(\theta) = \abs{P(e^{i\theta}),
+    #     H = I - T,
+    #     H = e^{-id\theta} R(e^{i\theta})
+    #     $$
+    #
+    # Here H and T are defined on reals, so the initial definition of R is only on the unit circle.
+    # We analytically continue this definition to the entire complex plane by replacing $e^{i\theta}$ by $z$.
+    R = Polynomial.basis(d) - Polynomial(P) * Polynomial(np.conj(P[::-1]))
+    roots = R.roots()
+
+    # R is self-inversive, so larger_roots and smaller_roots occur in conjugate pairs.
+    units: list[complex] = []  # roots r s.t. \abs{r} = 1
+    larger_roots: list[complex] = []  # roots r s.t. \abs{r} > 1
+    smaller_roots: list[complex] = []  # roots r s.t. \abs{r} < 1
+
+    for r in roots:
+        if np.allclose(np.abs(r), 1):
+            units.append(r)
+        elif np.abs(r) > 1:
+            larger_roots.append(r)
+        else:
+            smaller_roots.append(r)
+
+    if verify:
+        # verify that the non-unit roots indeed occur in conjugate pairs.
+        def is_permutation(A, B):
+            assert len(A) == len(B)
+            A = list(A)
+            for z in B:
+                for w in A:
+                    if np.allclose(z, w):
+                        A.remove(w)
+                        break
+                else:
+                    return False
+            return True
+
+        assert is_permutation(smaller_roots, 1 / np.array(larger_roots).conj())
+
+    # pair up roots in `units`, claimed in Eq. 40 and the explanation preceding it.
+    # all unit roots must have even multiplicity.
+    paired_units: list[complex] = []
+    unpaired_units: list[complex] = []
+    for z in units:
+        matched_z = None
+        for w in unpaired_units:
+            if np.allclose(z, w):
+                matched_z = w
+                break
+
+        if matched_z:
+            paired_units.append(z)
+            unpaired_units.remove(matched_z)
+        else:
+            unpaired_units.append(z)
+
+    if verify:
+        assert len(unpaired_units) == 0
+
+    # Q = G \hat{G}, where
+    # - \hat{G}^2 is the monomials which are unit roots of R, which occur in pairs.
+    # - G*(z) G(z) is the interpolation of the conjugate paired non-unit roots of R,
+    #   described in Eq. 37 - Eq. 38
+
+    # Leading co-efficient of R described in Eq. 37.
+    # Note: In the paper, the polynomial is interpolated from larger_roots,
+    #       but this is swapped in our implementation to reduce the error in Q.
+    c = R.coef[-1]
+    scaling_factor = np.sqrt(np.abs(c * np.prod(larger_roots)))
+
+    Q = scaling_factor * Polynomial.fromroots(paired_units + smaller_roots)
+
+    return np.around(Q.coef, decimals=10)
```

Comment:
Why the magic number `10` ? Does this have any impact on potential loss of precision?

Reply:
actually this isn't needed, I removed it now.

But the current tests only work up till degree 4, the errors from degree 5 are quite large already. I plan to look into this next.

Reply:
>  the errors from degree 5 are quite large already. I plan to look into this next.

So swapping smaller_roots and larger_roots doesn't help? Or it works till degree 4 because of reduced error due to swapping the use of two ?

Reply:
Btw, the complementary polynomial satisfies the required constraint $|P(z)|^2 + |Q(z)|^2 = 1$ even for degree 100, so the error seems to blow-up in the QSP phase computation.

Reply:
@tanujkhattar thoughts?

Reply:
Looks good, happy to merge now and debug the precision issue in follow up PRs.